File: console/lib/checklist.ts
// TS port of agents/shared/checklist.py validate_publish_checklist
// Validates against live Compendium taxonomy (data/taxonomy/*.json)

import { METHODOLOGY_CODES, SPECIALTY_OPTIONS } from "@/lib/taxonomy"

const LEGACY_METHODOLOGY_PREFIXES = ["RS-", "OD-", "EI-", "QI-"] as const
const LIVE_METHODOLOGY_CODES = new Set(METHODOLOGY_CODES)

const METHODOLOGY_OPTIONAL_TYPES = new Set([
  "software",
  "community",
  "funding",
  "dataset",
  "template",
  "visual_reference",
])

function methodologyRequiredForType(resourceType: string | undefined): boolean {
  if (!resourceType) return true
  return !METHODOLOGY_OPTIONAL_TYPES.has(resourceType)
}

const VOCAB_SPECIALTY_CODES = new Set(SPECIALTY_OPTIONS.map((s) => s.code))

function normalizeMethodologyCode(code: string): string {
  return code.trim().toUpperCase()
}

function normalizeSpecialtyCode(value: string): string | null {
  const raw = value.trim()
  if (!raw) return null
  const upper = raw.toUpperCase()
  if (VOCAB_SPECIALTY_CODES.has(upper)) return upper
  const slug = raw.toLowerCase()
  const match = SPECIALTY_OPTIONS.find((s) => s.slug === slug)
  return match?.code ?? null
}

export interface ChecklistError {
  field: string
  message: string
}

export function validatePublishChecklist(
  record: Record<string, unknown>,
  reviewedBy: string
): ChecklistError[] {
  const errors: ChecklistError[] = []

  // 1. editorial_description present
  const desc = record.editorial_description as string | undefined
  if (!desc || desc.trim().length === 0) {
    errors.push({ field: "editorial_description", message: "Editorial description is required" })
  }

  // 2. ≥1 live platform methodology code (type-aware)
  const codes = (record.methodology_codes as string[] | undefined) || []
  const typeCode = record.resource_type_code as string | undefined
  if (methodologyRequiredForType(typeCode) && codes.length === 0) {
    errors.push({
      field: "methodology_codes",
      message: "At least one platform methodology code required",
    })
  } else if (codes.length > 0) {
    for (const code of codes) {
      for (const legacy of LEGACY_METHODOLOGY_PREFIXES) {
        if (code.startsWith(legacy)) {
          errors.push({
            field: "methodology_codes",
            message: `Legacy methodology code '${code}' — emit platform codes only`,
          })
        }
      }
      const norm = normalizeMethodologyCode(code)
      if (!LIVE_METHODOLOGY_CODES.has(norm)) {
        errors.push({
          field: "methodology_codes",
          message: `Unknown platform code '${code}' (not in live Compendium taxonomy)`,
        })
      }
    }
  }

  // 2b. discipline_codes — validate specialty codes when present (optional field)
  const disciplines = record.discipline_codes as string[] | undefined
  if (disciplines && disciplines.length > 0) {
    for (const raw of disciplines) {
      const norm = normalizeSpecialtyCode(raw)
      if (!norm) {
        errors.push({
          field: "discipline_codes",
          message: `Invalid specialty code '${raw}' (not in vocabulary)`,
        })
      }
    }
  }

  // 3. quality_score ≥ 60
  const qs = record.quality_score as number | undefined
  if (qs === undefined || qs === null) {
    errors.push({ field: "quality_score", message: "quality_score is required" })
  } else if (qs < 60) {
    errors.push({ field: "quality_score", message: `quality_score ${qs} is below minimum (60)` })
  }

  // 4. URL present
  const url = record.url as string | undefined
  if (!url || url.trim().length === 0) {
    errors.push({ field: "url", message: "Resource URL is required" })
  }

  // 5 & 6. editorial_reviewed_by (injected by caller)
  if (!reviewedBy || reviewedBy.trim().length === 0) {
    errors.push({ field: "editorial_reviewed_by", message: "Reviewer identity required" })
  }

  return errors
}

File: agents/shared/checklist.py
"""
Publish checklist — validates a draft record before it can be set to `published`.

From docs/SCHEMA.md:
  editorial_description present
  ≥1 methodology_code (platform)
  quality_score present and ≥ 60
  url present (link verified)
  human-ratified (editorial_reviewed_by + editorial_reviewed_at)

Returns a list of error strings; empty list = passes.
"""
from __future__ import annotations

from agents.shared.codes import LEGACY_METHODOLOGY_PREFIXES
from agents.taxonomy import is_valid_methodology_code, normalize_methodology_code


def validate_publish_checklist(record: dict) -> list[str]:
    """
    Validate a draft record against the publish checklist.
    Returns a list of violation strings. Empty = publishable.
    """
    errors: list[str] = []

    # 1. editorial_description present
    if not record.get("editorial_description", "").strip():
        errors.append("editorial_description is missing or empty")

    # 2. ≥1 platform methodology_code
    codes = record.get("methodology_codes", [])
    if not codes:
        errors.append("methodology_codes: ≥1 platform code required (SYN-/OBS-/EVAL-…)")
    else:
        for code in codes:
            for legacy in LEGACY_METHODOLOGY_PREFIXES:
                if code.startswith(legacy):
                    errors.append(
                        f"Legacy methodology code {code!r} (prefix {legacy!r}) in "
                        f"methodology_codes — emit platform codes only"
                    )
            norm = normalize_methodology_code(code)
            if not is_valid_methodology_code(norm):
                errors.append(
                    f"methodology_codes: unknown platform code {code!r} "
                    f"(not in live Compendium taxonomy)"
                )

    # 3. quality_score present and ≥ 60
    quality_score = record.get("quality_score")
    if quality_score is None:
        errors.append("quality_score is missing")
    elif quality_score < 60:
        errors.append(
            f"quality_score={quality_score} is below 60 — "
            "records scoring below 60 are not publishable (hidden on card)"
        )

    # 4. url present (link verified)
    if not record.get("url", "").strip():
        errors.append("url is missing — link must be verified before publishing")

    # 5. Human ratification: editorial_reviewed_by + editorial_reviewed_at
    if not record.get("editorial_reviewed_by"):
        errors.append(
            "editorial_reviewed_by is missing — record must be human-ratified before publishing"
        )
    if not record.get("editorial_reviewed_at"):
        errors.append(
            "editorial_reviewed_at is missing — record must be human-ratified before publishing"
        )

    return errors
